File: packages/b2c-tooling-sdk/src/operations/ods/wait-for-sandbox.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */

import type {OdsClient} from '../../clients/ods.js';
import {getLogger} from '../../logging/logger.js';

/**
 * State of a sandbox in the ODS system.
 *
 * Standard states:
 * - `deleted` - Sandbox has been deleted
 * - `failed` - Sandbox creation or operation failed
 * - `started` - Sandbox is running
 * - `stopped` - Sandbox is stopped
 *
 * Also accepts any other string value via `(string & {})` for forward compatibility
 * with future ODS API states.
 */
export type SandboxState = 'deleted' | 'failed' | 'started' | 'stopped' | (string & {});

/**
 * Error thrown when a sandbox does not reach its target state within the specified timeout period.
 *
 * @param sandboxId - ID of the sandbox being monitored
 * @param targetState - The desired state that was not reached before timeout
 * @param timeoutSeconds - Timeout duration in seconds that was exceeded
 * @param lastState - Optional last observed state before timeout occurred
 */
export class SandboxPollingTimeoutError extends Error {
  constructor(
    public readonly sandboxId: string,
    public readonly targetState: SandboxState,
    public readonly timeoutSeconds: number,
    public readonly lastState?: SandboxState,
  ) {
    super(
      `Timeout waiting for sandbox ${sandboxId} to reach state ${targetState} after ${timeoutSeconds} seconds${
        lastState ? ` (lastState=${lastState})` : ''
      }`,
    );
    this.name = 'SandboxPollingTimeoutError';
  }
}

/**
 * Error thrown when fetching sandbox status fails during polling.
 *
 * @param sandboxId - ID of the sandbox being polled
 * @param message - Error message or details from the API response
 */
export class SandboxPollingError extends Error {
  constructor(
    public readonly sandboxId: string,
    public readonly message: string,
  ) {
    super(`Failed to fetch sandbox status for ${sandboxId}: ${message}`);
    this.name = 'SandboxPollingError';
  }
}

/**
 * Error thrown when a sandbox reaches a terminal error state (failed or deleted) during polling.
 *
 * @param sandboxId - ID of the sandbox that reached a terminal state
 * @param targetState - The desired state that was being waited for
 * @param state - The terminal state that was reached
 */
export class SandboxTerminalStateError extends Error {
  constructor(
    public readonly sandboxId: string,
    public readonly targetState: SandboxState,
    public readonly state: SandboxState,
  ) {
    super(`Sandbox ${sandboxId} reached terminal state ${state} while waiting for ${targetState}`);
    this.name = 'SandboxTerminalStateError';
  }
}

/**
 * Information provided to the `onPoll` callback on each poll during sandbox state monitoring.
 *
 * @property sandboxId - ID of the sandbox being monitored
 * @property elapsedSeconds - Total elapsed time in seconds since polling started
 * @property state - Current observed state of the sandbox
 */
export interface WaitForSandboxPollInfo {
  sandboxId: string;
  elapsedSeconds: number;
  state: SandboxState;
}

/**
 * Configuration options for {@link waitForSandbox} sandbox state polling.
 *
 * @property sandboxId - ID of the sandbox to monitor
 * @property targetState - Desired sandbox state to wait for
 * @property pollIntervalSeconds - Interval between status checks in seconds
 * @property timeoutSeconds - Maximum time to wait in seconds (0 = no timeout)
 * @property onPoll - Optional callback invoked on each poll with current state
 * @property sleep - Optional custom sleep function (primarily for testing)
 */
export interface WaitForSandboxOptions {
  sandboxId: string;
  targetState: SandboxState;
  pollIntervalSeconds: number;
  timeoutSeconds: number;
  onPoll?: (info: WaitForSandboxPollInfo) => void;
  sleep?: (ms: number) => Promise<void>;
}

async function defaultSleep(ms: number): Promise<void> {
  await new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

/**
 * Waits for a sandbox to reach a target state by polling its status.
 *
 * This function polls the ODS API at regular intervals until the sandbox reaches the desired state,
 * times out, or enters a terminal error state (failed/deleted).
 *
 * @param client - ODS client for API calls
 * @param options - Polling configuration options
 * @param options.sandboxId - ID of the sandbox to monitor
 * @param options.targetState - Desired sandbox state (e.g., 'started', 'stopped', 'deleted')
 * @param options.pollIntervalSeconds - Seconds between status checks
 * @param options.timeoutSeconds - Maximum seconds to wait (0 = no timeout)
 * @param options.onPoll - Optional callback invoked on each poll with current state
 * @param options.sleep - Optional custom sleep function (primarily for testing)
 *
 * @throws {SandboxPollingTimeoutError} If the timeout is exceeded before reaching target state
 * @throws {SandboxPollingError} If the API request fails
 * @throws {SandboxTerminalStateError} If the sandbox enters a terminal error state
 *
 * @example
 * ```typescript
 * await waitForSandbox(odsClient, {
 *   sandboxId: 'abc-123',
 *   targetState: 'started',
 *   pollIntervalSeconds: 5,
 *   timeoutSeconds: 300,
 *   onPoll: (info) => console.log(`State: ${info.state} (${info.elapsedSeconds}s)`),
 * });
 * ```
 */
export async function waitForSandbox(client: OdsClient, options: WaitForSandboxOptions): Promise<void> {
  const logger = getLogger();
  const {sandboxId, targetState, pollIntervalSeconds, timeoutSeconds} = options;

  const sleepFn = options.sleep ?? defaultSleep;
  const startTime = Date.now();
  const pollIntervalMs = pollIntervalSeconds * 1000;
  const timeoutMs = timeoutSeconds * 1000;

  await sleepFn(pollIntervalMs);

  let lastState: SandboxState | undefined;

  while (true) {
    const elapsedSeconds = Math.round((Date.now() - startTime) / 1000);

    if (timeoutSeconds > 0 && Date.now() - startTime > timeoutMs) {
      throw new SandboxPollingTimeoutError(sandboxId, targetState, timeoutSeconds, lastState);
    }

    const result = await client.GET('/sandboxes/{sandboxId}', {
      params: {
        path: {sandboxId},
      },
    });

    if (!result.data?.data) {
      throw new SandboxPollingError(sandboxId, result.response?.statusText || 'Unknown error');
    }

    const sandbox = result.data.data;
    const currentState = (sandbox.state as SandboxState) || 'unknown';
    lastState = currentState;

    logger.trace({sandboxId, elapsedSeconds, state: currentState}, '[ODS] Sandbox poll');
    options.onPoll?.({sandboxId, elapsedSeconds, state: currentState});

    if (currentState === targetState) {
      return;
    }

    if (currentState === 'failed' || currentState === 'deleted') {
      throw new SandboxTerminalStateError(sandboxId, targetState, currentState);
    }

    await sleepFn(pollIntervalMs);
  }
}

File: python/b2c-tooling-sdk/src/b2c_tooling_sdk/operations/ods/wait_for_sandbox.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""Polls a sandbox's status until it reaches a target state.

Mirrors ``src/operations/ods/wait-for-sandbox.ts``.
"""

from __future__ import annotations

import asyncio
import time
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Any

from b2c_tooling_sdk.clients import OdsClient
from b2c_tooling_sdk.logging import get_logger

#: State of a sandbox in the ODS system.
#:
#: Standard states:
#:
#: - ``deleted`` - Sandbox has been deleted
#: - ``failed`` - Sandbox creation or operation failed
#: - ``started`` - Sandbox is running
#: - ``stopped`` - Sandbox is stopped
#:
#: Also accepts any other string value for forward compatibility with future
#: ODS API states (mirrors the TS ``'deleted' | 'failed' | 'started' | 'stopped'
#: | (string & {})``).
SandboxState = str

#: Sandbox states that terminate polling with an error.
_TERMINAL_STATES = ("failed", "deleted")


class SandboxPollingTimeoutError(Exception):
    """Raised when a sandbox does not reach its target state within the specified timeout period."""

    def __init__(
        self,
        sandbox_id: str,
        target_state: SandboxState,
        timeout_seconds: float,
        last_state: SandboxState | None = None,
    ) -> None:
        suffix = f" (lastState={last_state})" if last_state else ""
        super().__init__(
            f"Timeout waiting for sandbox {sandbox_id} to reach state {target_state} "
            f"after {timeout_seconds} seconds{suffix}"
        )
        self.name = "SandboxPollingTimeoutError"
        self.sandbox_id = sandbox_id
        self.target_state = target_state
        self.timeout_seconds = timeout_seconds
        self.last_state = last_state


class SandboxPollingError(Exception):
    """Raised when fetching sandbox status fails during polling."""

    def __init__(self, sandbox_id: str, message: str) -> None:
        super().__init__(f"Failed to fetch sandbox status for {sandbox_id}: {message}")
        self.name = "SandboxPollingError"
        self.sandbox_id = sandbox_id
        self.message = message


class SandboxTerminalStateError(Exception):
    """Raised when a sandbox reaches a terminal error state (failed or deleted) during polling."""

    def __init__(self, sandbox_id: str, target_state: SandboxState, state: SandboxState) -> None:
        super().__init__(f"Sandbox {sandbox_id} reached terminal state {state} while waiting for {target_state}")
        self.name = "SandboxTerminalStateError"
        self.sandbox_id = sandbox_id
        self.target_state = target_state
        self.state = state


@dataclass
class WaitForSandboxPollInfo:
    """Information provided to the ``on_poll`` callback on each poll during sandbox state monitoring."""

    sandbox_id: str
    elapsed_seconds: float
    state: SandboxState


async def _default_sleep(seconds: float) -> None:
    await asyncio.sleep(seconds)


@dataclass
class WaitForSandboxOptions:
    """Configuration options for :func:`wait_for_sandbox` sandbox state polling."""

    sandbox_id: str
    target_state: SandboxState
    poll_interval_seconds: float
    timeout_seconds: float
    #: Optional callback invoked on each poll with current state.
    on_poll: Callable[[WaitForSandboxPollInfo], None] | None = None
    #: Optional custom sleep function (primarily for testing). Receives the sleep
    #: duration in seconds.
    sleep: Callable[[float], Awaitable[None]] = _default_sleep


async def wait_for_sandbox(client: OdsClient, options: WaitForSandboxOptions) -> None:
    """Wait for a sandbox to reach a target state by polling its status.

    Polls the ODS API at regular intervals until the sandbox reaches the desired
    state, times out, or enters a terminal error state (failed/deleted).

    :param client: ODS client for API calls.
    :param options: Polling configuration options.
    :raises SandboxPollingTimeoutError: If the timeout is exceeded before reaching target state.
    :raises SandboxPollingError: If the API request fails.
    :raises SandboxTerminalStateError: If the sandbox enters a terminal error state.

    :example:
        >>> await wait_for_sandbox(
        ...     ods_client,
        ...     WaitForSandboxOptions(
        ...         sandbox_id="abc-123",
        ...         target_state="started",
        ...         poll_interval_seconds=5,
        ...         timeout_seconds=300,
        ...         on_poll=lambda info: print(f"State: {info.state} ({info.elapsed_seconds}s)"),
        ...     ),
        ... )
    """
    logger = get_logger("operations.ods")
    sandbox_id = options.sandbox_id
    target_state = options.target_state
    poll_interval_seconds = options.poll_interval_seconds
    timeout_seconds = options.timeout_seconds

    sleep_fn = options.sleep
    start_time = time.monotonic()

    await sleep_fn(poll_interval_seconds)

    last_state: SandboxState | None = None

    while True:
        elapsed_seconds = round(time.monotonic() - start_time)

        if timeout_seconds > 0 and time.monotonic() - start_time > timeout_seconds:
            raise SandboxPollingTimeoutError(sandbox_id, target_state, timeout_seconds, last_state)

        result = await client.get("/sandboxes/{sandboxId}", {"params": {"path": {"sandboxId": sandbox_id}}})

        sandbox: dict[str, Any] | None = result.data.get("data") if isinstance(result.data, dict) else None
        if not sandbox:
            reason = result.response.reason_phrase if result.response is not None else ""
            raise SandboxPollingError(sandbox_id, reason or "Unknown error")

        current_state: SandboxState = sandbox.get("state") or "unknown"
        last_state = current_state

        logger.debug(
            "[ODS] Sandbox poll: sandbox_id=%s elapsed_seconds=%s state=%s",
            sandbox_id,
            elapsed_seconds,
            current_state,
        )
        if options.on_poll is not None:
            options.on_poll(
                WaitForSandboxPollInfo(sandbox_id=sandbox_id, elapsed_seconds=elapsed_seconds, state=current_state)
            )

        if current_state == target_state:
            return

        if current_state in _TERMINAL_STATES:
            raise SandboxTerminalStateError(sandbox_id, target_state, current_state)

        await sleep_fn(poll_interval_seconds)


__all__ = [
    "SandboxPollingError",
    "SandboxPollingTimeoutError",
    "SandboxState",
    "SandboxTerminalStateError",
    "WaitForSandboxOptions",
    "WaitForSandboxPollInfo",
    "wait_for_sandbox",
]
